### PR TITLE
Validate ProwJobs using CustomResourceValidation

### LINKED ISSUE #4373: Validate ProwJobs using CRD validation
https://github.com/kubernetes/community/blob/31af349b1777e0fead2dd42eccca5799fb0375a2/contributors/design-proposals/api-machinery/customresources-validation.md

Things we need to validate:
- [ ] status.type and status.state are known values
- [ ] avoid validating the agent so we can support out of tree agent controllers
- [ ] completion time is set on a pj with status.state == {success, failure, error, aborted}
- [ ] children jobs have the same pull+sha as parents for presubmits
- [ ] max_concurrency is non-negative

I will keep this list updated as more things pop up.

1.8 thing
/area prow

### PR DESCRIPTION
**Pull request by nikhita**
Fixes #4373 

- [x] status.type and status.state are known values
- [x] avoid validating the agent so we can support out of tree agent controllers
- [x] completion time is set on a pj with status.state == {success, failure, error, aborted}
- [x] max_concurrency is non-negative
- [ ] children jobs have the same pull+sha as parents for presubmits

/area prow
/cc kargakis sttts spxtr

### CODE REVIEW

**Comment by nikhita** [comment]
> children jobs have the same pull+sha as parents for presubmits

For this one, I believe it is not possible to validate via JSON Schema. JSON Schema does not provide constructs to compare if two fields have the same values.

Unless I am missing something. @sttts?

File: prow/architecture.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ type PluginClient struct {
...
+The [trigger](https://github.com/kubernetes/test-infra/tree/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/trigger) plugin has a function called [`handleIC`](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/trigger/ic.go#L31) that it [registered at init-time](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/plugins/trigger/trigger.go#L35). This function checks the issue comment for any test requests by comparing the issue comment body to the list of jobs supplied in the `Config` object in the `PluginClient`. I requested that some tests run, so the trigger plugin talks to the Kubernetes API server and creates a [ProwJob Custom Resource](https://github.com/kubernetes/test-infra/blob/c8829eef589a044126289cb5b4dc8e85db3ea22f/prow/kube/prowjob.go#L50-L83) with the information from the issue comment. At this point, `cmd/hook`'s role is done. It goes on and handles the next webhooks.
```
Comment (nikhita) [inline prow/architecture.md]: Just changed `ThirdPartryResource` -> `Custom Resource`.

**Comment by kargakis** [comment]
Awesome! Looks good to me but needs a review from @cjwagner too
/lgtm
/hold

@nikhita once we merge this, mind adding your changes in https://github.com/openshift/release/blob/master/cluster/ci/config/prow/prow_crd.yaml too?

**Comment by nikhita** [comment]
> @nikhita once we merge this, mind adding your changes in https://github.com/openshift/release/blob/master/cluster/ci/config/prow/prow_crd.yaml too?

Will do! :)

**Comment by sttts** [comment]
> For this one, I believe it is not possible to validate via JSON Schema. JSON Schema does not provide constructs to compare if two fields have the same values.
> Unless I am missing something. @sttts?

Without having seen the types, cross-field checks are not possible.

File: prow/cluster/prow_job.yaml [hunk 2]
```diff
@@ -10,3 +10,31 @@ spec:
     singular: prowjob
     plural: prowjobs
   scope: Namespaced
+  validation:
+      openAPIV3Schema:
```
Comment (spxtr) [inline prow/cluster/prow_job.yaml]: Nit: two spaces rather than 4

File: prow/cluster/prow_job.yaml [hunk 3]
```diff
@@ -10,3 +10,31 @@ spec:
...
+                properties:
+                  state:
+                    type: string
+                    enum:
+                    - "success"
+                    - "failure"
+                    - "error"
+                    - "aborted"
+            - required:
+              - completionTime
```
Comment (spxtr) [inline prow/cluster/prow_job.yaml]: Does this mean that this is required all the time? Running prowjobs don't have a completion time, but it's probably set to the zero value somewhere.

Comment (cjwagner) [reply prow/cluster/prow_job.yaml]: I think the `anyOf` operator handles this. Either `completionTime` must be specified or `state` must not be "success", "failure", "error", or "aborted".

File: prow/cluster/prow_job.yaml [hunk 4]
```diff
@@ -10,3 +10,31 @@ spec:
...
+      openAPIV3Schema:
+        properties:
+          spec:
+            properties:
+              max_concurrency:
+                type: integer
+                minimum: 0
+              type:
+                type: string
+                pattern: '^(presubmit|postsubmit|periodic|batch)$'
```
Comment (cjwagner) [inline prow/cluster/prow_job.yaml]: These patterns could be `enum`s.

File: prow/cluster/prow_job.yaml (hunk 2 above)
Comment (nikhita) [reply prow/cluster/prow_job.yaml]: Done.

File: prow/cluster/prow_job.yaml (hunk 4 above)
Comment (nikhita) [reply prow/cluster/prow_job.yaml]: Done.

File: prow/cluster/prow_job.yaml (hunk 3 above)
Comment (nikhita) [reply prow/cluster/prow_job.yaml]: > I think the anyOf operator handles this. Either completionTime must be specified or state must not be "success", "failure", "error", or "aborted".

Yes, this is what happens.

Comment (kargakis) [reply prow/cluster/prow_job.yaml]: @spxtr are you worried that running prowjobs may be invalidated? It should be easy to verify it and we should probably stop setting it to the zero value.

Comment (kargakis) [reply prow/cluster/prow_job.yaml]: Hrm, the issue is probably with Go defaulting. Any arguments against turning [CompletionTime](https://github.com/kubernetes/test-infra/blob/f72668786f9002196e04337728b3c7a19ecacdd3/prow/kube/prowjob.go#L96) into a pointer?

Comment (kargakis) [reply prow/cluster/prow_job.yaml]: Opened https://github.com/kubernetes/test-infra/pull/6215

Comment (cjwagner) [reply prow/cluster/prow_job.yaml]: Why is there a problem with this? Shouldn't the `omitempty` annotation prevent the `CompletionTime` field from being included if it is 0?

Comment (kargakis) [reply prow/cluster/prow_job.yaml]: Go structs do not work with `omitempty` as expected: https://stackoverflow.com/questions/32643815/golang-json-omitempty-with-time-time-field/32646035#32646035

Comment (kargakis) [reply prow/cluster/prow_job.yaml]: Also, you can verify this today by doing a `kubectl get prowjob <name> -o yaml` on a pending/triggered prowjob. `CompletionTime` is set (ie. not dropped by `omitempty`).

**Comment by nikhita** [comment]
Can this be merged, now that https://github.com/kubernetes/test-infra/pull/6215 is in?

**Comment by BenTheElder** [comment]
Thanks for this! 😄 
/hold cancel

**Pull request merged**